File: sdks/go/autonoma/teardown.go
package autonoma

import (
	"context"
	"fmt"
	"sort"
	"strings"
)

// TeardownOrderInfo holds the precomputed teardown order for a schema.
type TeardownOrderInfo struct {
	Order            []string
	ScopeRootModel   string
	Cycles           [][]string
	ScopeFieldByModel map[string]string
}

// ComputeTeardownOrder computes the deletion order for models (reverse topological order).
func ComputeTeardownOrder(schema SchemaInfo) TeardownOrderInfo {
	var scopeRootModel string
	for _, edge := range schema.Edges {
		if strings.EqualFold(edge.LocalField, schema.ScopeField) && edge.To != edge.From {
			scopeRootModel = edge.To
			break
		}
	}

	scopeFieldByModel := make(map[string]string)
	if scopeRootModel != "" {
		for _, edge := range schema.Edges {
			if edge.To == scopeRootModel && edge.From != scopeRootModel {
				scopeFieldByModel[edge.From] = edge.LocalField
			}
		}
	}

	modelNames := make([]string, len(schema.Models))
	for i, m := range schema.Models {
		modelNames[i] = m.Name
	}

	result := TopoSort(modelNames, schema.Edges)

	// Build condensation graph
	var components [][]string
	nodeToComp := make(map[string]int)

	for _, cycle := range result.Cycles {
		idx := len(components)
		components = append(components, cycle)
		for _, node := range cycle {
			nodeToComp[node] = idx
		}
	}
	for _, node := range result.Sorted {
		nodeToComp[node] = len(components)
		components = append(components, []string{node})
	}

	compCount := len(components)
	condAdj := make([]map[int]bool, compCount)
	condInDeg := make([]int, compCount)
	for i := 0; i < compCount; i++ {
		condAdj[i] = make(map[int]bool)
	}
	for _, edge := range schema.Edges {
		if edge.From == edge.To {
			continue
		}
		fc, fcOk := nodeToComp[edge.From]
		tc, tcOk := nodeToComp[edge.To]
		if fcOk && tcOk && fc != tc && !condAdj[tc][fc] {
			condAdj[tc][fc] = true
			condInDeg[fc]++
		}
	}

	var condQueue []int
	for i := 0; i < compCount; i++ {
		if condInDeg[i] == 0 {
			condQueue = append(condQueue, i)
		}
	}
	sort.Ints(condQueue)
	var condOrder []int
	for len(condQueue) > 0 {
		sort.Ints(condQueue)
		idx := condQueue[0]
		condQueue = condQueue[1:]
		condOrder = append(condOrder, idx)
		for neighbor := range condAdj[idx] {
			condInDeg[neighbor]--
			if condInDeg[neighbor] == 0 {
				condQueue = append(condQueue, neighbor)
			}
		}
	}

	// Flatten in reverse condensation order, excluding scope root
	var order []string
	for i := len(condOrder) - 1; i >= 0; i-- {
		for _, model := range components[condOrder[i]] {
			if model != scopeRootModel {
				order = append(order, model)
			}
		}
	}

	return TeardownOrderInfo{
		Order:            order,
		ScopeRootModel:   scopeRootModel,
		Cycles:           result.Cycles,
		ScopeFieldByModel: scopeFieldByModel,
	}
}

// Teardown deletes all data scoped to a value, in reverse topological order.
// skipModels contains model names that should be excluded from SQL deletion
// (because they were already handled by factory teardown).
func Teardown(
	ctx context.Context,
	executor SQLExecutor,
	dialect *Dialect,
	tableMap map[string]string,
	columnMaps map[string]map[string]string,
	schema SchemaInfo,
	scopeValue string,
	refs map[string][]map[string]any,
	skipModels map[string]bool,
) error {
	tdInfo := ComputeTeardownOrder(schema)

	return executor.Transaction(ctx, func(tx SQLExecutor) error {
		// Break cycles by nullifying deferrable FKs
		for _, cycle := range tdInfo.Cycles {
			edge := FindDeferrableEdge(cycle, schema.Edges)
			if edge != nil {
				scopeFK, hasScopeFK := tdInfo.ScopeFieldByModel[edge.From]
				if hasScopeFK {
					dbTable := tableMap[edge.From]
					colMap := columnMaps[edge.From]
					if colMap == nil {
						colMap = make(map[string]string)
					}
					if dbTable != "" {
						dbFKCol := colMap[edge.LocalField]
						if dbFKCol == "" {
							dbFKCol = edge.LocalField
						}
						dbScopeCol := colMap[scopeFK]
						if dbScopeCol == "" {
							dbScopeCol = scopeFK
						}
						sql := fmt.Sprintf("UPDATE %s SET %s = NULL WHERE %s = %s",
							dialect.QuoteID(dbTable),
							dialect.QuoteID(dbFKCol),
							dialect.QuoteID(dbScopeCol),
							dialect.Param(1))
						if _, err := tx.Query(ctx, sql, scopeValue); err != nil {
							return err
						}
					}
				}
			}
		}

		// Delete in order (dependents first), skipping factory-teardown models
		for _, model := range tdInfo.Order {
			if skipModels[model] {
				continue
			}
			if err := deleteModel(ctx, tx, dialect, tableMap, columnMaps, model, scopeValue, tdInfo.ScopeFieldByModel, refs, schema); err != nil {
				return err
			}
		}

		// Delete scope root last (unless skipped by factory teardown)
		if tdInfo.ScopeRootModel == "" || skipModels[tdInfo.ScopeRootModel] {
			return nil
		}

		dbTable := tableMap[tdInfo.ScopeRootModel]
		colMap := columnMaps[tdInfo.ScopeRootModel]
		if colMap == nil {
			colMap = make(map[string]string)
		}
		if dbTable != "" {
			rootPkFieldName := "id"
			for _, mi := range schema.Models {
				if mi.Name == tdInfo.ScopeRootModel {
					var firstId string
					for _, f := range mi.Fields {
						if f.IsId {
							if firstId == "" {
								firstId = f.Name
							}
							if strings.EqualFold(f.Name, "id") {
								rootPkFieldName = f.Name
								firstId = ""
								break
							}
						}
					}
					if firstId != "" {
						rootPkFieldName = firstId
					}
					break
				}
			}
			idCol := colMap[rootPkFieldName]
			if idCol == "" {
				idCol = rootPkFieldName
			}
			sql := fmt.Sprintf("DELETE FROM %s WHERE %s = %s",
				dialect.QuoteID(dbTable),
				dialect.QuoteID(idCol),
				dialect.Param(1))
			if _, err := tx.Query(ctx, sql, scopeValue); err != nil {
				return err
			}
		}

		return nil
	})
}

func deleteModel(
	ctx context.Context,
	tx SQLExecutor,
	dialect *Dialect,
	tableMap map[string]string,
	columnMaps map[string]map[string]string,
	model string,
	scopeValue string,
	scopeFieldByModel map[string]string,
	refs map[string][]map[string]any,
	schema SchemaInfo,
) error {
	dbTable := tableMap[model]
	if dbTable == "" {
		return nil
	}
	colMap := columnMaps[model]
	if colMap == nil {
		colMap = make(map[string]string)
	}

	// Bug 4: find actual PK field name from schema
	// When multiple IsId fields exist (composite PK), prefer the one named "id"
	pkFieldName := "id"
	for _, mi := range schema.Models {
		if mi.Name == model {
			var firstId string
			for _, f := range mi.Fields {
				if f.IsId {
					if firstId == "" {
						firstId = f.Name
					}
					if strings.EqualFold(f.Name, "id") {
						pkFieldName = f.Name
						firstId = ""
						break
					}
				}
			}
			if firstId != "" {
				pkFieldName = firstId
			}
			break
		}
	}

	scopeFK, hasScopeFK := scopeFieldByModel[model]
	if hasScopeFK {
		dbCol := colMap[scopeFK]
		if dbCol == "" {
			dbCol = scopeFK
		}
		sql := fmt.Sprintf("DELETE FROM %s WHERE %s = %s",
			dialect.QuoteID(dbTable),
			dialect.QuoteID(dbCol),
			dialect.Param(1))
		_, err := tx.Query(ctx, sql, scopeValue)
		return err
	}

	if refs != nil {
		records := refs[model]
		if len(records) > 0 {
			// Bug 3: accept any non-nil value, not just strings
			var ids []any
			for _, r := range records {
				if id, ok := r[pkFieldName]; ok && id != nil {
					ids = append(ids, id)
				}
			}
			if len(ids) > 0 {
				idCol := colMap[pkFieldName]
				if idCol == "" {
					idCol = pkFieldName
				}
				var placeholders []string
				for i := range ids {
					placeholders = append(placeholders, dialect.Param(i+1))
				}
				sql := fmt.Sprintf("DELETE FROM %s WHERE %s IN (%s)",
					dialect.QuoteID(dbTable),
					dialect.QuoteID(idCol),
					strings.Join(placeholders, ", "))
				_, err := tx.Query(ctx, sql, ids...)
				return err
			}
		}
	}

	return nil
}

File: sdks/typescript/packages/sdk-prisma/src/teardown.ts
import type { SQLExecutor, SchemaInfo } from './types'
import type { Dialect } from './dialect'
import { topoSort, findDeferrableEdge } from './graph'

/**
 * Tear down all data scoped to a value, in reverse topological order.
 *
 * Strategy:
 *   1. Find the scope root model (e.g. Organization) from FK edges
 *   2. Any model with a FK pointing to the scope root is "scoped"
 *   3. Delete scoped models by their FK = scopeValue
 *   4. Delete non-scoped models by their record IDs from refs
 *   5. Delete the scope root entity last by id = scopeValue
 */
export async function teardown(
  executor: SQLExecutor,
  dialect: Dialect,
  tableMap: Map<string, string>,
  columnMaps: Map<string, Map<string, string>>,
  schema: SchemaInfo,
  scopeValue: string,
  refs?: Record<string, Record<string, unknown>[]>,
): Promise<void> {
  // Find scope root: the model that the scopeField FK points TO
  let scopeRootModel: string | null = null
  for (const edge of schema.edges) {
    if (edge.localField.toLowerCase() === schema.scopeField.toLowerCase() && edge.to !== edge.from) {
      scopeRootModel = edge.to
      break
    }
  }

  // Build map: model → FK field name that points to the scope root
  const scopeFieldByModel = new Map<string, string>()
  if (scopeRootModel) {
    for (const edge of schema.edges) {
      if (edge.to === scopeRootModel && edge.from !== scopeRootModel) {
        scopeFieldByModel.set(edge.from, edge.localField)
      }
    }
  }

  const modelNames = schema.models.map((m) => m.name)
  const { sorted, cycles } = topoSort(modelNames, schema.edges)

  await executor.transaction(async (tx) => {
    // Break cycles by nullifying deferrable FKs
    for (const cycle of cycles) {
      const edge = findDeferrableEdge(cycle, schema.edges)
      if (edge) {
        const scopeFK = scopeFieldByModel.get(edge.from)
        if (scopeFK) {
          const dbTable = tableMap.get(edge.from)
          const colMap = columnMaps.get(edge.from) ?? new Map<string, string>()
          if (dbTable) {
            const dbFKCol = colMap.get(edge.localField) ?? edge.localField
            const dbScopeCol = colMap.get(scopeFK) ?? scopeFK
            await tx.query(
              `UPDATE ${dialect.quoteId(dbTable)} SET ${dialect.quoteId(dbFKCol)} = NULL WHERE ${dialect.quoteId(dbScopeCol)} = ${dialect.param(1)}`,
              [scopeValue],
            )
          }
        }
      }
    }

    // Build condensation graph: each SCC is a super-node, each sorted node
    // is its own node. Topo-sort the condensation DAG and delete in reverse
    // order so that dependents of cycles are deleted before the cycle itself.
    const components: string[][] = []
    const nodeToComp = new Map<string, number>()

    for (const cycle of cycles) {
      const idx = components.length
      components.push(cycle)
      for (const node of cycle) nodeToComp.set(node, idx)
    }
    for (const node of sorted) {
      nodeToComp.set(node, components.length)
      components.push([node])
    }

    // Build condensation DAG edges (dependency → dependent)
    const condAdj = new Map<number, Set<number>>()
    const condInDeg = new Map<number, number>()
    for (let i = 0; i < components.length; i++) {
      condAdj.set(i, new Set())
      condInDeg.set(i, 0)
    }
    for (const edge of schema.edges) {
      if (edge.from === edge.to) continue
      const fc = nodeToComp.get(edge.from)
      const tc = nodeToComp.get(edge.to)
      if (fc !== undefined && tc !== undefined && fc !== tc && !condAdj.get(tc)!.has(fc)) {
        condAdj.get(tc)!.add(fc)
        condInDeg.set(fc, (condInDeg.get(fc) ?? 0) + 1)
      }
    }

    // Kahn's algorithm on the condensation DAG
    const condQueue: number[] = []
    for (const [idx, deg] of condInDeg) {
      if (deg === 0) condQueue.push(idx)
    }
    const condOrder: number[] = []
    while (condQueue.length > 0) {
      condQueue.sort()
      const idx = condQueue.shift()!
      condOrder.push(idx)
      for (const neighbor of condAdj.get(idx)!) {
        const nd = (condInDeg.get(neighbor) ?? 1) - 1
        condInDeg.set(neighbor, nd)
        if (nd === 0) condQueue.push(neighbor)
      }
    }

    // Delete in reverse condensation order (dependents first)
    for (const compIdx of [...condOrder].reverse()) {
      for (const model of components[compIdx]) {
        if (model === scopeRootModel) continue
        await deleteModel(tx, dialect, tableMap, columnMaps, model, scopeValue, scopeFieldByModel, refs, schema)
      }
    }

    // Delete the scope root entity last
    if (scopeRootModel) {
      const dbTable = tableMap.get(scopeRootModel)
      const colMap = columnMaps.get(scopeRootModel) ?? new Map<string, string>()
      if (dbTable) {
        const rootModelInfo = schema.models.find((m) => m.name === scopeRootModel)
        const rootIdFields = rootModelInfo?.fields.filter((f) => f.isId) ?? []
        const rootPkFieldName = (rootIdFields.find((f) => f.name.toLowerCase() === 'id') ?? rootIdFields[0])?.name ?? 'id'
        const idCol = colMap.get(rootPkFieldName) ?? rootPkFieldName
        await tx.query(
          `DELETE FROM ${dialect.quoteId(dbTable)} WHERE ${dialect.quoteId(idCol)} = ${dialect.param(1)}`,
          [scopeValue],
        )
      }
    }
  })
}

async function deleteModel(
  tx: SQLExecutor,
  dialect: Dialect,
  tableMap: Map<string, string>,
  columnMaps: Map<string, Map<string, string>>,
  model: string,
  scopeValue: string,
  scopeFieldByModel: Map<string, string>,
  refs: Record<string, Record<string, unknown>[]> | undefined,
  schema: SchemaInfo,
): Promise<void> {
  const dbTable = tableMap.get(model)
  if (!dbTable) return
  const colMap = columnMaps.get(model) ?? new Map<string, string>()

  // Find actual PK field name from schema
  const modelInfo = schema.models.find((m) => m.name === model)
  const idFields = modelInfo?.fields.filter((f) => f.isId) ?? []
  const pkFieldName = (idFields.find((f) => f.name.toLowerCase() === 'id') ?? idFields[0])?.name ?? 'id'

  const scopeFK = scopeFieldByModel.get(model)
  if (scopeFK) {
    // Has FK to scope root → delete by that FK
    const dbCol = colMap.get(scopeFK) ?? scopeFK
    await tx.query(
      `DELETE FROM ${dialect.quoteId(dbTable)} WHERE ${dialect.quoteId(dbCol)} = ${dialect.param(1)}`,
      [scopeValue],
    )
  } else if (refs?.[model]) {
    // No FK to scope root, but we created records → delete by IDs
    const ids = refs[model]
      .map((r) => r[pkFieldName])
      .filter((id): id is string | number => id != null)
    if (ids.length > 0) {
      const idCol = colMap.get(pkFieldName) ?? pkFieldName
      const placeholders = ids.map((_, i) => dialect.param(i + 1)).join(', ')
      await tx.query(
        `DELETE FROM ${dialect.quoteId(dbTable)} WHERE ${dialect.quoteId(idCol)} IN (${placeholders})`,
        ids,
      )
    }
  }
}
